C/wrc20.c

```diff
@@ -0,0 +1,133 @@
+  b += (a & 0xff00000000000000)>>56;
+  b += (a & 0x00ff000000000000)>>40;
+  b += (a & 0x0000ff0000000000)>>24;
+  b += (a & 0x000000ff00000000)>>8;
+  b += (a & 0x00000000ff000000)<<8;
+  b += (a & 0x0000000000ff0000)<<24;
+  b += (a & 0x000000000000ff00)<<40;
+  b += (a & 0x00000000000000ff)<<56;
+  return b;
+}
+
+
+// global data used in next function, will be allocated to WebAssembly memory
+bytes32 addy[1] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
+bytes32 balance[1] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
+
+void do_balance() {
+  if (getCallDataSize() != 24)
+    revert(0, 0);
+
+  // get address to check balance of, note: padded to 32 bytes since used as key
+  callDataCopy((i32ptr*)addy, 4, 20);
+
+  // get balance
+  storageLoad((i32ptr*)addy, (i32ptr*)balance);
+
+  // return balance
+  finish((i32ptr*)balance, 32);
+}
+
+
+// global data used in next function, will be allocated to WebAssembly memory
+bytes32 sender[1] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
+bytes32 recipient[1] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
+bytes32 value[1] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
+bytes32 recipient_balance[1] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
+bytes32 sender_balance[1] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
+
+void do_transfer() {
+  if (getCallDataSize() != 32)
+    revert(0, 0);
+
+  // get caller
+  getCaller((i32ptr*)sender);
+
+  // get recipient from message at byte 4, length 20
+  callDataCopy((i32ptr*)recipient, 4, 20);
+
+  // get amount to transfer from message at byte 24, length 8
+  callDataCopy((i32ptr*)value, 24, 8);
+  *(i64*)value = reverse_bytes(*(i64*)value);
+
+  // get balances
+  storageLoad((i32ptr*)sender, (i32ptr*)sender_balance);
+  storageLoad((i32ptr*)recipient, (i32ptr*)recipient_balance);
+  *(i64*)sender_balance = reverse_bytes(*(i64*)sender_balance);
+  *(i64*)recipient_balance = reverse_bytes(*(i64*)recipient_balance);
+
+  // make sure sender has enough
+  if (*(i64*)sender_balance < *(i64*)value)
+    revert(0, 0);
+
+  // adjust balances
+  * (i64*)sender_balance -= * (i64*)value;
+  * (i64*)recipient_balance += * (i64*)value;
+
+  // store results
+  *(i64*)sender_balance = reverse_bytes(*(i64*)sender_balance);
+  *(i64*)recipient_balance = reverse_bytes(*(i64*)recipient_balance);
+  storageStore((i32ptr*)sender, (i32ptr*)sender_balance);
+  storageStore((i32ptr*)recipient, (i32ptr*)recipient_balance);
+
+}
+
+
+// global data used in next function, will be allocated to WebAssembly memory
+i32 selector[1] = {0};
+
+void _main(void) {
+  if (getCallDataSize() < 4)
+    revert(0, 0);
+
+  // get first four bytes of message, use for calling
+  callDataCopy((i32ptr*)selector, 0, 4); //from byte 0, length 4
+
+  // call function based on selector value
+  switch (*selector) {
+    case 0x1a029399:
```

Comment:
I'm curious, why are the selector values here different from those in the example pseudo-code for this repository?

Reply:
I think that you are referring to the wrc20 spec https://gist.github.com/axic/16158c5c88fbc7b1d09dfa8c658bc363 where the selector has value `0x9993021a`. The selector here is reversed is because `callDataCopy()` puts the call data into wasm memory, and Wasm reverses the bytes when it loads/stores them to the stack. I could have also done `reverse_bytes()` on selector and used the same selector byte-ordering as the spec.